packages/react/src/hooks/contracts/useContractRead.ts

```diff
@@ -121,40 +121,12 @@ export function useContractRead({
     return enabled
   }, [addressOrName, blockNumber, cacheOnBlock, enabled_, functionName])
 
-  const client = useQueryClient()
-  React.useEffect(() => {
-    if (enabled) {
-      const unwatch = watchReadContract(
-        {
-          addressOrName,
-          args,
-          chainId,
-          contractInterface,
-          functionName,
-          overrides,
-          listenToBlock: watch && !cacheOnBlock,
-        },
-        (result) => client.setQueryData(queryKey_, result),
-      )
-      return unwatch
-    }
-  }, [
-    addressOrName,
-    args,
-    cacheOnBlock,
-    chainId,
-    client,
-    contractInterface,
-    enabled,
-    functionName,
-    overrides,
-    queryKey_,
-    watch,
-  ])
+  useInvalidateOnBlock({ enabled: watch && !cacheOnBlock, queryKey: queryKey_ })
```

Comment:
The old implementation was also refetching on `provider` changes, but we are already checking that in our query key (via chain id). I have refactored this to "invalidate on new block" if `watch & !cacheOnBlock` is truthy.